eclair-core/src/main/scala/fr/acinq/eclair/payment/Bolt12Invoice.scala

```diff
@@ -0,0 +1,164 @@
+
+  private val paths: Option[Seq[RouteBlinding.BlindedRoute]] = records.get[Paths].map(_.paths)
+
+  private val blindedpay: Option[Seq[PayInfo]] = records.get[BlindedPay].map(_.payInfos)
+
+  private val blindedCapacities: Option[Seq[MilliSatoshi]] = records.get[BlindedCapacities].map(_.capacities)
+
+  require(paths.map(_.map(_.blindedNodes.length - 1).sum) == blindedpay.map(_.length))
+  require(blindedCapacities.forall(paths.map(_.length) contains _.length))
+
+  val blindedPaths: Option[Seq[BlindedPath]] = paths.map(routes => {
+    var remainingPayInfos = blindedpay.get
+    routes.zip(blindedCapacities.map(_.map(Some(_))).getOrElse(routes.map(_ => None))).map {
+      case (route, capacity) =>
+        val payInfos = remainingPayInfos.take(route.blindedNodes.length - 1)
+        remainingPayInfos = remainingPayInfos.drop(route.blindedNodes.length - 1)
+        BlindedPath(route, payInfos, capacity)
+    }
+  })
+
+  val issuer: Option[String] = records.get[Issuer].map(_.issuer)
+
+  val quantity: Option[Long] = records.get[Quantity].map(_.quantity)
+
+  val refundFor: Option[ByteVector32] = records.get[RefundFor].map(_.refundedPaymentHash)
+
+  val payerKey: Option[ByteVector32] = records.get[PayerKey].map(_.key)
+
+  val payerNote: Option[String] = records.get[PayerNote].map(_.note)
+
+  val payerInfo: Option[ByteVector] = records.get[PayerInfo].map(_.info)
+
+  val fallbacks: Option[Seq[FallbackAddress]] = records.get[Fallbacks].map(_.addresses)
+
+  val refundSignature: Option[ByteVector64] = records.get[RefundSignature].map(_.signature)
+
+  val replaceInvoice: Option[ByteVector32] = records.get[ReplaceInvoice].map(_.paymentHash)
+
+  val signature: ByteVector64 = records.get[Signature].get.signature
+
+  def isValidFor(offer: Offer, request: InvoiceRequest): Boolean = {
+    nodeId.value.drop(1) == offer.nodeId.value.drop(1) &&
+      checkSignature &&
+      offerId.contains(offer.offerId) &&
+      offer.chains.contains(chain) &&
+      !isExpired() &&
+      request.amount.contains(amount) &&
+      quantity == request.quantity_opt &&
+      payerKey.contains(request.payerKey) &&
+      payerInfo == request.payerInfo &&
+      payerNote.forall(request.payerNote.contains(_)) &&
+      description == Left(offer.description) &&
+      issuer == offer.issuer &&
+      request.features.areSupported(features)
+  }
+
+  def checkRefundSignature: Boolean = {
+    (refundSignature, refundFor, payerKey) match {
+      case (Some(sig), Some(hash), Some(key)) =>
+        verifySchnorr("lightning" + "invoice" + "payer_signature", hash, sig, key)
+      case _ => false
+    }
+  }
+
+  def checkSignature: Boolean = {
+    val withoutSig = TlvStream(records.records.filter { case _: Signature => false case _ => true }, records.unknown)
+    verifySchnorr("lightning" + "invoice" + "signature", rootHash(withoutSig, invoiceTlvCodec).get, signature, ByteVector32(nodeId.value.drop(1)))
+  }
+
+  def withNodeId(id: PublicKey): Bolt12Invoice =
```

Comment:
Interesting, it's hard to judge whether it makes sense as-is or if there may be a clearer way of doing that.
Can you remove it for now and re-add it in the offers PR? It will be easier to review with the code that uses it as it's not just a spec type.